src/psyclone/psyir/transformations/arrayassignment2loops_trans.py

```diff
@@ -368,6 +367,34 @@ def validate(self, node, options=None):
                 # pylint: disable=cell-var-from-loop
                 raise TransformationError(LazyString(
                     lambda: f"{message} In:\n{node.debug_string()}"))
+            # We must understand any array accesses on the RHS. This means all
+            # array indices must be known (because otherwise they themselves
+            # might be arrays, e.g.
+            #      a(:) = b(c) where `c` can be a scalar or an array.
+            if isinstance(reference, ArrayMixin):
```

Comment:
Again, do we still have a problem if the ArrayMixin is not in the top-level reference? e.g.
`a(:) = b%field(c)`
(maybe it should walk for inner arraymixins?)